src/lib/salad/hll.c

```diff
@@ -0,0 +1,223 @@
+#include "hll.h"
+#include "hll_empirical.h"
+
+#include <trivia/util.h>
+
+#include <stddef.h>
+#include <math.h>
+#include <assert.h>
+
+/* 6 bits are needed to store the number of leading zeros of 64 bit hash */
+#define HLL_REGISTER_BITS  6
+
+static inline size_t
+hll_n_registers(uint8_t precision)
+{
+	return 1U << precision;
+}
+
+static inline int
+hll_is_valid_cache(double cached)
```

Comment:
Why not pass `struct hll` here?

Reply:
I thought there is no need in `hll` here. I'l pass `const struct hll *` to make it more understandable.